prov/efa/src/efa_strerror.c

```diff
@@ -101,18 +128,92 @@ static const char *efa_errno_str(enum efa_errno err)
 }
 
+static const char *efa_debug_str(enum efa_errno err)
+{
+	switch (err) {
+	case FI_EFA_REMOTE_ERROR_BAD_DEST_QPN:
+		return "This error can be caused by a crashed peer. "
+			   "Please verify the receiver status";
+	case FI_EFA_LOCAL_ERROR_UNRESP_REMOTE:
+		return "This error can be caused by a crashed peer or "
+			   "incorrect inbound/outbound rules in the security group - "
+			   "EFA requires \"All traffic\" type allowlisting. "
+			   "Please also verify the receiver status.";
+	default:
+		return NULL;
+	}
+}
+
 /**
- * @brief   Convert EFA error into a printable string
- * @details Given a non-negative EFA-specific error code, this function returns
- *          a pointer to a static string that corresponds to it. This text is
- *          suitable for printing in debug/warning/error messages.
+ * @brief Write connection information to an error message string
+ * 
+ * @param[in]	err_data	efa_error_data pointer that contains the connection information
+ * @param[out]	err_str		The destination string
+ * @return		A status code. 0 if successful, otherwise a negative integer.
+*/
+int write_efa_err_conn_str(const struct efa_error_data *err_data, char *err_str)
+{
+	if (!err_data || !err_str)
+		return -FI_EINVAL;
+
+	int ret = snprintf(err_str, CONN_INFO_MAX_LENGTH, CONN_INFO_FORMAT,
+					   this_or_default(err_data->ep_raw_addr_str, NA_STR),
+					   this_or_default(efa_host_id_str(err_data->ep_host_id), NA_STR),
+					   this_or_default(err_data->peer_raw_addr_str, NA_STR),
+					   this_or_default(efa_host_id_str(err_data->peer_host_id), NA_STR));
+
+	if (ret < 0)
+		return -FI_EOTHER;
+
+	if (ret >= CONN_INFO_MAX_LENGTH)
+		return -FI_ENOBUFS;
+
+	return 0;
+}
+
+/**
+ * @brief	Convert EFA error into a printable string
+ * @details	Given a non-negative EFA-specific error code and optional
+ *			efa_error_data, this function returns a pointer to a static
+ *			string that corresponds to it. This text is suitable for
+ *			printing in debug/warning/error messages.
  *
- * @param [in] err An EFA-specific error code
- * @return     Pointer to a string corresponding to err
+ * @param[in]	err			An EFA-specific error code
+ * @param[in]	err_data	Pointer to struct efa_error_data
+ * @return		An error message with helpful debugging information specific to
+ *				the error code
  */
-const char *efa_strerror(enum efa_errno err)
+const char *efa_strerror(enum efa_errno err, const struct efa_error_data *err_data)
 {
-	return err >= FI_EFA_ERRNO_OFFSET
-		? efa_errno_str(err)
-		: efa_io_comp_status_str(err);
+	const char *base_err = err >= FI_EFA_ERRNO_OFFSET
+							   ? efa_errno_str(err)
+							   : efa_io_comp_status_str(err);
+	const char *debug_hint = efa_debug_str(err);
+	size_t err_str_len = strlen(base_err);
+
+	/* Separate string with a newline */
+	err_str_len += debug_hint ? strlen(debug_hint) + 1 : 0;
+	err_str_len += err_data ? CONN_INFO_MAX_LENGTH + 1 : 0;
+	err_str_len += 1; /* \0 */
+
+	char *err_str = malloc(err_str_len);
```

Comment:
Each call to this function will malloc a piece of memory, which is totally unnecessary.